File: GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/views/QuizActivity.java
package com.bignerdranch.android.geoquiz.views;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.Gravity;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import com.bignerdranch.android.geoquiz.BuildConfig;
import com.bignerdranch.android.geoquiz.controllers.QuestionController;
import com.bignerdranch.android.geoquiz.models.Question;
import com.bignerdranch.android.geoquiz.R;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import androidx.annotation.StringRes;
import timber.log.Timber;

public class QuizActivity extends AppCompatActivity {
  private static final String TAG = "QuizActivity";
  private static final String KEY_INDEX = "index";

  @SuppressWarnings("FieldCanBeLocal")
  private Button trueButton;
  @SuppressWarnings("FieldCanBeLocal")
  private Button falseButton;
  private ImageButton nextButton;
  private ImageButton prevButton;
  private TextView questionTextView;
  private List<Question> questionBank = QuestionController.getInstance().getQuestionBank();
  private int currentIndex = 0;
  private int countOfCorrectQuestionsResolved = 0;
  private int countOfQuestionsResolved = 0;
  @NonNull
  private final Set<Question> resolvedQuestions = new HashSet();

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    if (BuildConfig.DEBUG) {
      Timber.plant(new Timber.DebugTree());
    }
    Timber.d("onCreate(Bundle) called");
    setContentView(R.layout.activity_quiz);

    if (savedInstanceState != null) {
      currentIndex = savedInstanceState.getInt(KEY_INDEX, 0);
    }

    questionTextView = findViewById(R.id.questionTextView);
    questionTextView.setOnClickListener(v -> {
      currentIndex = (currentIndex + 1) % questionBank.size();
      updateQuestion();
    });

    trueButton = findViewById(R.id.questionButtonTrue);
    trueButton.setOnClickListener(v -> updateView(true));

    falseButton = findViewById(R.id.questionButtonFalse);
    falseButton.setOnClickListener(v -> updateView(false));

    prevButton = findViewById(R.id.prevButton);
    prevButton.setOnClickListener(v -> {
      if (currentIndex > 0) {
        currentIndex = (currentIndex - 1) % questionBank.size();
        updateQuestion();
      }
    });
    nextButton = findViewById(R.id.nextButton);
    nextButton.setOnClickListener(v -> {
      currentIndex = (currentIndex + 1) % questionBank.size();
      updateQuestion();
    });

    updateQuestion();
  }

  @Override
  public void onSaveInstanceState(Bundle savedInstanceState) {
    super.onSaveInstanceState(savedInstanceState);
    Timber.d("onSaveInstanceState");
    savedInstanceState.putInt(KEY_INDEX, currentIndex);
  }

  @Override
  public void onStart() {
    super.onStart();
    Timber.d("onStart() called");
  }

  @Override
  public void onPause() {
    super.onPause();
    Timber.d("onPause() called");
  }

  @Override
  public void onResume() {
    super.onResume();
    Timber.d("onResume() called");
  }

  @Override
  public void onStop() {
    super.onStop();
    Timber.d("onStop() called");
  }

  @Override
  public void onDestroy() {
    super.onDestroy();
    Timber.d("onDestroy() called");
  }

  private void updateQuestion() {
    Question actualQuestion = questionBank.get(currentIndex);
    questionTextView.setText(actualQuestion.getTextResId());
    setButtonsState(!(resolvedQuestions.contains(actualQuestion)));
    if (countOfQuestionsResolved == questionBank.size()) {
      String percentage = (countOfCorrectQuestionsResolved) + "/" + (countOfQuestionsResolved);
      Toast.makeText(this, "The average is: " + percentage, Toast.LENGTH_SHORT).show();
    }
  }

  private void updateView(boolean isCorrect) {
    Question actualQuestion = questionBank.get(currentIndex);
    if (resolvedQuestions.contains(actualQuestion)) {
      setButtonsState(false);
    } else {
      showToastCorrectly(isCorrect);
      increaseCounters(isCorrect);
      updateResolvedQuestions(actualQuestion);
    }
  }

  private void showToastCorrectly(boolean isCorrect) {
    @StringRes
    int answerTest = isCorrect ? R.string.correct : R.string.incorrect;
    Toast toast = Toast.makeText(this, answerTest, Toast.LENGTH_SHORT);
    toast.setGravity(Gravity.TOP | Gravity.CENTER_HORIZONTAL, 0, 0);
    toast.show();
  }

  private void increaseCounters(boolean isCorrect) {
    if (isCorrect) {
      countOfCorrectQuestionsResolved += 1;
    }
    countOfQuestionsResolved += 1;
  }

  @NonNull
  private void updateResolvedQuestions(Question actualQuestion) {
    resolvedQuestions.add(actualQuestion);
    setButtonsState(false);
  }

  private void setButtonsState(boolean enabled) {
    falseButton.setEnabled(enabled);
    trueButton.setEnabled(enabled);
  }
}

File: app/src/main/java/com/bignerdranch/android/geoquiz/views/QuizActivity.kt
package com.bignerdranch.android.geoquiz.views

import androidx.appcompat.app.AppCompatActivity

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.Gravity
import android.widget.Toast

import com.bignerdranch.android.geoquiz.BuildConfig
import com.bignerdranch.android.geoquiz.models.Question
import com.bignerdranch.android.geoquiz.R

import java.util.HashSet

import androidx.annotation.StringRes
import androidx.lifecycle.ViewModelProviders
import com.bignerdranch.android.geoquiz.models.QuizViewModel
import kotlinx.android.synthetic.main.activity_quiz.*

import timber.log.Timber

class QuizActivity : AppCompatActivity() {
    companion object {
        private val KEY_INDEX = "index"
        private val REQUEST_CODE_CHEAT = 0
    }

    private var countOfCorrectQuestionsResolved = 0
    private var countOfQuestionsResolved = 0
    private var isCheater: Boolean = false
    private val resolvedQuestions = HashSet<Question>()
    private val quizViewModel: QuizViewModel by lazy {
        ViewModelProviders.of(this).get(QuizViewModel::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (BuildConfig.DEBUG) {
            Timber.plant(Timber.DebugTree())
        }
        Timber.d("onCreate(Bundle) called")
        setContentView(R.layout.activity_quiz)

        val currentIndex = savedInstanceState?.getInt(KEY_INDEX, 0) ?: 0
        quizViewModel.currentIndex = currentIndex

        questionTextView.setOnClickListener { v ->
            quizViewModel.moveToNext()
            updateQuestion()
        }

        questionButtonTrue.setOnClickListener { v -> updateView(true) }

        questionButtonFalse.setOnClickListener { v -> updateView(false) }

        cheatButton.setOnClickListener { v ->
            val answerIsTrue = quizViewModel.currentQuestionAnswer
            val intent = CheatActivity.newIntent(this, answerIsTrue)
            startActivityForResult(intent, REQUEST_CODE_CHEAT)
        }

        prevButton.setOnClickListener { v ->
            if (quizViewModel.currentIndex > 0) {
                quizViewModel.moveToPrev()
                isCheater = false
                updateQuestion()
            }
        }
        nextButton.setOnClickListener { v ->
            quizViewModel.moveToNext()
            isCheater = false
            updateQuestion()
        }

        updateQuestion()
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode != Activity.RESULT_OK) {
            return
        }

        if (requestCode == REQUEST_CODE_CHEAT) {
            if (data == null) {
                return
            }
            isCheater = CheatActivity.wasAnswerShown(data)
        }
    }

    public override fun onSaveInstanceState(savedInstanceState: Bundle) {
        super.onSaveInstanceState(savedInstanceState)
        Timber.d("onSaveInstanceState")
        savedInstanceState.putInt(KEY_INDEX, quizViewModel.currentIndex)
    }

    public override fun onStart() {
        super.onStart()
        Timber.d("onStart() called")
    }

    public override fun onPause() {
        super.onPause()
        Timber.d("onPause() called")
    }

    public override fun onResume() {
        super.onResume()
        Timber.d("onResume() called")
    }

    public override fun onStop() {
        super.onStop()
        Timber.d("onStop() called")
    }

    public override fun onDestroy() {
        super.onDestroy()
        Timber.d("onDestroy() called")
    }

    private fun updateQuestion() {
        val actualQuestion = quizViewModel.currentQuestion
        questionTextView.setText(quizViewModel.currentQuestionText)
        setButtonsState(!resolvedQuestions.contains(actualQuestion))
        showPercentage()
    }

    private fun updateView(isCorrect: Boolean) {
        val actualQuestion = quizViewModel.currentQuestion
        if (resolvedQuestions.contains(actualQuestion)) {
            setButtonsState(false)
        } else {
            showToastCorrectly(isCorrect)
            increaseCounters(isCorrect)
            updateResolvedQuestions(actualQuestion)
        }
    }

    private fun showToastCorrectly(isCorrect: Boolean) {
        @StringRes
        val answerTest = if (isCheater) R.string.judgment_toast else getAnswerText(isCorrect)
        val toast = Toast.makeText(this, answerTest, Toast.LENGTH_SHORT)
        toast.setGravity(Gravity.TOP or Gravity.CENTER_HORIZONTAL, 0, 0)
        toast.show()
    }

    @StringRes
    private fun getAnswerText(isCorrect: Boolean): Int =
            if (isCorrect) R.string.correct else R.string.incorrect

    private fun increaseCounters(isCorrect: Boolean) {
        if (isCorrect) {
            countOfCorrectQuestionsResolved += 1
        }
        countOfQuestionsResolved += 1
    }

    private fun updateResolvedQuestions(actualQuestion: Question) {
        resolvedQuestions.add(actualQuestion)
        setButtonsState(false)
    }

    private fun setButtonsState(enabled: Boolean) {
        questionButtonFalse!!.isEnabled = enabled
        questionButtonTrue!!.isEnabled = enabled
    }

    private fun showPercentage() {
        if (countOfQuestionsResolved == quizViewModel.questionBank.size) {
            val percentage = "$countOfCorrectQuestionsResolved/$countOfQuestionsResolved"
            Toast.makeText(this, "The average is: $percentage", Toast.LENGTH_SHORT).show()
        }
    }

}
